flink-streaming-java/src/main/java/org/apache/flink/streaming/api/windowing/triggers/ContinuousProcessingTimeTrigger.java

```diff
@@ -74,6 +74,11 @@ public TriggerResult onEventTime(long time, W window, TriggerContext ctx) throws
     @Override
     public TriggerResult onProcessingTime(long time, W window, TriggerContext ctx)
             throws Exception {
+
+        if (time == window.maxTimestamp()) {
```

Comment:
Is this guaranteed to work? I haven't found any indication that there is always a call `onProcessingTime(window.maxTimestamp(), window, ctx)`?

Maybe it would work consistently with `long nextFireTimestamp = Math.min(start + interval, window.maxTimestamp());`. Then state, timer state, and this method are actually in sync and we wouldn't need this special case.

Reply:
Sorry for the late reply. When processElement is called, it will register a CleanupTimer, which will use window.maxTimestamp(). When this time is triggered（onProcessingTime(window.maxTimestamp(), window, ctx),  it does not fire the window, and the state was cleared anyway.(org.apache.flink.streaming.runtime.operators.windowing.WindowOperator#registerCleanupTimer)

 I think `long nextFireTimestamp = Math.min(start + interval, window.maxTimestamp());` is reasonable, and make window.max window easier to understand. But in this way, we may have to modify several places, such as ContinuousProcessingTimeTrigger#onElement #onProcessingTime. In addition, ContinuousEventTimeTrigger is  also determines whether to fire the window through `time == window.maxTimestamp()` at first, modify it together?
https://github.com/apache/flink/blob/ac203a7a26d78fe270bd9941bb721be53a893040/flink-streaming-java/src/main/java/org/apache/flink/streaming/api/windowing/triggers/ContinuousEventTimeTrigger.java#L76